### PR TITLE
Attaching a message to an issue links to GitHub account, not MM account

### PR DESCRIPTION
**Pull request by abdulsmapara**
#### Summary
* Attaching a message to an issue links to GitHub account, not MM account
   - Instead of using the MM username, the username on GitHub is used. 
   - If the user whose message is attached is not connected to GitHub, then the MM username is used
* Added permalink to Message in reply, which is given after a user attaches a message to a GitHub issue

#### Ticket Link

  Fixes https://github.com/mattermost/mattermost-plugin-github/issues/253

### CODE REVIEW

File: server/api.go [hunk 1]
```diff
@@ -727,23 +727,31 @@ func (p *Plugin) createIssueComment(w http.ResponseWriter, r *http.Request, user
 		return
 	}
 
+	commentUsername := ""
 	commentUser, appErr := api.GetUser(post.UserId)
 	if appErr != nil {
 		writeAPIError(w, &APIErrorResponse{ID: "", Message: "failed to load post.UserID " + post.UserId + ": not found", StatusCode: http.StatusInternalServerError})
 		return
 	}
```
Comment (hanzei) [inline server/api.go:735]: Does this need to be called when `p.getGitHubUserInfo(post.UserId)` succeeds?

File: server/api.go [hunk 2]
```diff
@@ -727,23 +727,31 @@ func (p *Plugin) createIssueComment(w http.ResponseWriter, r *http.Request, user
...
+	if info, err := p.getGitHubUserInfo(post.UserId); err != nil {
```
Comment (hanzei) [inline server/api.go:736]: While the logic is correct here, only a subset of errors returned by `getGitHubUserInfo` indicate that the user is not connected. I think it's valuable that these get logged. You can check the error type with 
```golang
	if info, apiEr := p.getGitHubUserInfo(post.UserId); apiEr != nil {
		if apiErr.ID == API_ERROR_ID_NOT_CONNECTED` {
```

File: server/api.go [hunk 3]
```diff
@@ -727,23 +727,31 @@ func (p *Plugin) createIssueComment(w http.ResponseWriter, r *http.Request, user
...
+	} else {
+		currentUsername = info.GitHubUsername
 	}
 
 	siteUrl := api.GetConfig().ServiceSettings.SiteURL
 
 	permalink := getPermaLink(*siteUrl, req.PostId, req.CurrentTeam)
 
-	permalinkMessage := fmt.Sprintf("*@%s attached a* [message](%s) *from @%s*\n", currentUser.Username, permalink, commentUser.Username)
+	permalinkMessage := fmt.Sprintf("*@%s attached a* [message](%s) *from @%s*\n", currentUsername, permalink, commentUsername)
```
Comment (hanzei) [inline server/api.go:754]: While you a touching this part of the code, I would like to think about a change unrelated to the original ticket: Does it make sense to reverse the order and post the original message first?

Before:
---
*@hanzei attached a* [message](0xc0002536a0/_redirect/pl/a3egsu1n9bd9jy98h9dhz3p4wo) *from @hanzei*
Some messsage

After:
--- 
Some message

*@hanzei attached a* [message](0xc0002536a0/_redirect/pl/a3egsu1n9bd9jy98h9dhz3p4wo) *from @hanzei*



Let me know what you think

**Comment by hanzei** [comment]
Another request: Could you link the word "Message" to the peramlink of the original post in https://github.com/mattermost/mattermost-plugin-github/pull/256/files#diff-0d5a44e9b96f048fa52b477ce6be6253R773?

File: server/api.go (hunk 2 above)
Comment (abdulsmapara) [reply server/api.go]: Done !

File: server/api.go (hunk 3 above)
Comment (abdulsmapara) [reply server/api.go:760]: In my opinion, the message which'll be attached might be long, so the info (@abdulsmapara attached a [message](https://github.com/mattermost/mattermost-plugin-github/pull/256) from @hanzei) might go unseen at the end of the attached message; hence, it is good to keep it above the attached message.
Still, if the change needs to be done, Iet me know.

**Comment by abdulsmapara** [comment]
> Another request: Could you link the word "Message" to the peramlink of the original post in https://github.com/mattermost/mattermost-plugin-github/pull/256/files#diff-0d5a44e9b96f048fa52b477ce6be6253R773?

I haven't modified the permalink. The word "message" is still linked to the permalink (of the original code).
Am I missing something here ?

**Comment by abdulsmapara** [comment]
> Another request: Could you link the word "Message" to the peramlink of the original post in https://github.com/mattermost/mattermost-plugin-github/pull/256/files#diff-0d5a44e9b96f048fa52b477ce6be6253R773?
```Go
permalinkReplyMessage := fmt.Sprintf("Message attached to [#%v](https://github.com/%v/%v/issues/%v)", req.Number, req.Owner, req.Repo, req.Number)
reply := &model.Post{
		Message:   permalinkReplyMessage ,
		ChannelId: post.ChannelId,
		RootId:    rootId,
		ParentId:  rootId,
		UserId:    userID,
}
```
Is this what is requested ?

**Comment by hanzei** [comment]
> > Another request: Could you link the word "Message" to the peramlink of the original post in [#256 (files)](https://github.com/mattermost/mattermost-plugin-github/pull/256/files#diff-0d5a44e9b96f048fa52b477ce6be6253R773)?
> 
> ```go
> permalinkReplyMessage := fmt.Sprintf("Message attached to [#%v](https://github.com/%v/%v/issues/%v)", req.Number, req.Owner, req.Repo, req.Number)
> reply := &model.Post{
> 		Message:   permalinkReplyMessage ,
> 		ChannelId: post.ChannelId,
> 		RootId:    rootId,
> 		ParentId:  rootId,
> 		UserId:    userID,
> }
> ```
> 
> Is this what is requested ?

Yes these are the lines in question.

File: server/api.go (hunk 3 above)
Comment (hanzei) [reply server/api.go:760]: I'm 0/5 on my suggestion. Let's leave it as it is for now.

**Comment by abdulsmapara** [comment]
> > > Another request: Could you link the word "Message" to the peramlink of the original post in [#256 (files)](https://github.com/mattermost/mattermost-plugin-github/pull/256/files#diff-0d5a44e9b96f048fa52b477ce6be6253R773)?
> > 
> > 
> > ```go
> > permalinkReplyMessage := fmt.Sprintf("Message attached to [#%v](https://github.com/%v/%v/issues/%v)", req.Number, req.Owner, req.Repo, req.Number)
> > reply := &model.Post{
> > 		Message:   permalinkReplyMessage ,
> > 		ChannelId: post.ChannelId,
> > 		RootId:    rootId,
> > 		ParentId:  rootId,
> > 		UserId:    userID,
> > }
> > ```
> > 
> > 
> > Is this what is requested ?
> 
> Yes these are the lines in question.

I haven't added a commit for this as I was not sure, should I add now @hanzei ?

**Comment by hanzei** [comment]
Please do so @abdulsmapara. It would be great if this change could be piggy backed in the PR.

**Comment by DHaussermann** [comment]
@hanzei If you have some time, could you please peer-test this PR using the  peer testing process documented here: https://mattermost.atlassian.net/wiki/spaces/INT/pages/619937892/Peer+Testing+Process

**Comment by hanzei** [comment]
@DHaussermann :point_up_2:

**Comment by hanzei** [comment]
@DHaussermann :point_up_2:

**Comment by hanzei** [comment]
@DHaussermann :point_up_2:

**Pull request merged**